presto-hive/src/test/java/com/facebook/presto/hive/TestBackgroundHiveSplitLoader.java

```diff
@@ -180,6 +181,105 @@ public void testEmptyFileWithNoBlocks()
     }
 
+    @Test
+    public void testNoHangIfPartitionIsOffline()
+            throws Exception
+    {
+        BackgroundHiveSplitLoader backgroundHiveSplitLoader = backgroundHiveSplitLoaderNoMetadata();
+        HiveSplitSource hiveSplitSource = hiveSplitSource(backgroundHiveSplitLoader, TupleDomain.none());
+        backgroundHiveSplitLoader.start(hiveSplitSource);
+
+        try {
+            drain(hiveSplitSource);
+        }
+        catch (Exception e) {
+            assertEquals(e.getMessage(), "java.lang.Exception: OFFLINE");
+        }
+
+        boolean failed = false;
+        try {
+            hiveSplitSource.isFinished();
+        }
+        catch (Exception e) {
+            failed = true;
+        }
+        assertEquals(failed, true);
+    }
+
+    private static HivePartitionMetadata verifyOnline(HivePartitionMetadata hivePartitionMetadata)
+            throws Exception
+    {
+        if (hivePartitionMetadata.getHivePartition().getTableName().getSchemaName().equalsIgnoreCase("testSchema2")) {
+            throw new Exception("OFFLINE");
+        }
+
+        return hivePartitionMetadata;
+    }
+
+    private static Iterable<HivePartitionMetadata> getIterableHivePartitionMetadata()
+            throws Exception
+    {
+        return new Iterable<HivePartitionMetadata>()
+        {
+            @Override
+            public Iterator<HivePartitionMetadata> iterator()
+            {
+                return new Iterator<HivePartitionMetadata>()
+                {
+                    private int position;
+                    private List<HivePartitionMetadata> hivePartitionMetadatas =
+                            ImmutableList.of(
+                                    new HivePartitionMetadata(
+                                            new HivePartition(new SchemaTableName("testSchema", "table_name"), ImmutableList.of()),
+                                            Optional.empty(),
+                                            ImmutableMap.of()),
+                                    new HivePartitionMetadata(
+                                            new HivePartition(new SchemaTableName("testSchema2", "table_name"), ImmutableList.of()),
+                                            Optional.empty(),
+                                            ImmutableMap.of()));
+
+                    @Override
+                    public boolean hasNext()
+                    {
+                        try {
+                            verifyOnline(hivePartitionMetadatas.get(position));
+                        }
+                        catch (Exception e) {
+                            throw new RuntimeException(e);
+                        }
+                        return true;
+                    }
+
+                    @Override
+                    public HivePartitionMetadata next()
+                    {
+                        return hivePartitionMetadatas.get(position++);
```

Comment:
violates `Iterator` contract, as it doesn't throw `NoSuchElementException` when iterator exhausted

At least extend `com.google.common.collect.AbstractIterator`, but better yet don't implement iterator at all:
```
hivePartitionMetadatas = ....
hivePartitionMetadatas.stream()
   .peek(this::verifyOnline)
   .iterator()
```

(a didn't read the test's logic, so won't comment on that)

Reply:
Thanks for the suggestion about the `AbstractIterator`, I'm updating the code in the next iteration to use it. For the test, I need some control on the `Iterator` as to make sure the code path is tested properly, a call to hasNext() will need to throw.

Reply:
> For the test, I need some control on the Iterator as to make sure the code path is tested properly, a call to hasNext() will need to throw.

Sounds like very tight coupling with the implementation you test.. which may change in the future, leaving the test potentially unsound.